cpp/src/arrow/compute/row/grouper.h

```diff
@@ -30,6 +30,49 @@
 namespace arrow {
 namespace compute {
 
+/// \brief A segment of contiguous rows for grouping
+struct ARROW_EXPORT GroupingSegment {
+  int64_t offset;
+  int64_t length;
+  bool is_open;
+  bool extends;
+};
+
+inline bool operator==(const GroupingSegment& segment1, const GroupingSegment& segment2) {
+  return segment1.offset == segment2.offset && segment1.length == segment2.length &&
+         segment1.is_open == segment2.is_open && segment1.extends == segment2.extends;
+}
+inline bool operator!=(const GroupingSegment& segment1, const GroupingSegment& segment2) {
+  return !(segment1 == segment2);
+}
+
+/// \brief Computes grouping segments for a batch. Each segment covers rows with identical
+/// values in the batch. The values in the batch are often selected as keys from a larger
+/// batch.
+class ARROW_EXPORT GroupingSegmenter {
+ public:
+  virtual ~GroupingSegmenter() = default;
+
+  /// \brief Construct a GroupingSegmenter which receives the specified key types
+  static Result<std::unique_ptr<GroupingSegmenter>> Make(
+      const std::vector<TypeHolder>& key_types, bool nullable_keys = false,
+      ExecContext* ctx = default_exec_context());
```

Comment:
Remove the default values for ctx and enforce it to be passed in explicitly?

Reply:
I feel defaults like this can lead to bugs, i.e., the caller has a ctx that is not default_exec_context() and forget to pass that in.

Reply:
This removal would break the [template-invocation discussed here](https://github.com/apache/arrow/pull/34311#discussion_r1116357888). I could refactor this invocation, if you feel strongly about the API here.

Reply:
I am not sure why this is related, can you elaborate

Reply:
The template-invocation `GroupClass::Make(...)` passes a single `const std::vector<TypeHolder>&` parameter. In one case it invokes `Grouper::Make` and in the second case `RowSegmenter::Make`. These doesn't have the same signature; in particular they differ on the default parameters. However, both accept a single `const std::vector<TypeHolder>&` parameter, which can be seen as an overload-signature that the invocation relies on. If the default values were removed, this overload-signature would be removed, and the invocation would break.

Reply:
So to clarify - the default values in the constructor here are because we want to template GroupClass and Grouper in test?

Reply:
I designed the default values to get a common interface before I had the specific template invocation in mind. If you feel strongly about removing the default values, one alternative is add functions in the tester file that would wrap the maker function with this common interface.

Reply:
In general I think having default values for internal function makes it easier to make mistakes such as forgetting to override a default value. Most of the "Make" function of ExecNode, for example, don't use default values.

"I designed the default values to get a common interface "

Can you elaborate this? To get a common interface for what and how is that related to having default values here?

Reply:
Oh I see - you are trying to create a "Make" interface for both `RowSegmenter` and `Grouper` class. IIUC, the common interface is only used in templating test since `RowSegmenter` and `Grouper` don't really share the same interface. Actually I found the templating of these two classes together quite confusing - all it is testing is the "Make" function. We might as well having them tested separately to avoid confusion.